testing/base/source_gen.cpp

```diff
@@ -0,0 +1,676 @@
+// distribution.
+//
+// Note that this template must be defined prior to its use below.
+template <typename AppendIds>
+auto SourceGen::GetIdsImpl(int number, int min_length, int max_length,
+                           bool uniform, AppendIds append_ids)
+    -> llvm::SmallVector<llvm::StringRef> {
+  CARBON_CHECK(min_length <= max_length);
+  CARBON_CHECK(uniform || max_length <= 64)
+      << "Cannot produce a meaningful non-uniform distribution of lengths "
+         "longer than 64 as those are exceedingly rare in our observed data "
+         "sets.";
+
+  llvm::SmallVector<llvm::StringRef> ids;
+  ids.reserve(number);
+
+  // First, compute how many identifiers of each size we'll need.
+  int count_sum =
+      uniform ? (max_length - min_length) + 1
+              : Sum(llvm::ArrayRef(IdLengthCounts)
+                        .slice(min_length - 1, max_length - min_length + 1));
+  CARBON_CHECK(count_sum >= 1);
+  int number_rem = number % count_sum;
+  for (int length : llvm::seq_inclusive(min_length, max_length)) {
+    // Scale this length if non-uniform.
+    int scale = uniform ? 1 : IdLengthCounts[length - 1];
+    int length_count = (number / count_sum) * scale;
+    if (number_rem > 0) {
+      length_count += std::min(scale, number_rem);
+      number_rem -= scale;
+    }
+    append_ids(length, length_count, ids);
+  }
+  CARBON_CHECK(static_cast<int>(ids.size()) == number);
+
+  return ids;
+}
+
+auto SourceGen::GetIds(int number, int min_length, int max_length, bool uniform)
+    -> llvm::SmallVector<llvm::StringRef> {
+  llvm::SmallVector<llvm::StringRef> ids =
+      GetIdsImpl(number, min_length, max_length, uniform,
+                 [this](int length, int length_count,
+                        llvm::SmallVectorImpl<llvm::StringRef>& dest) {
+                   auto length_ids = GetSingleLengthIds(length, length_count);
+                   dest.append(length_ids.begin(), length_ids.end());
+                 });
+
+  return ids;
+}
+
+auto SourceGen::GetUniqueIds(int number, int min_length, int max_length,
+                             bool uniform)
+    -> llvm::SmallVector<llvm::StringRef> {
+  CARBON_CHECK(min_length >= 4)
+      << "Cannot trivially guarantee enough distinct, unique identifiers for "
+         "lengths <= 3";
+  llvm::SmallVector<llvm::StringRef> ids =
+      GetIdsImpl(number, min_length, max_length, uniform,
+                 [this](int length, int length_count,
+                        llvm::SmallVectorImpl<llvm::StringRef>& dest) {
+                   AppendUniqueIdentifiers(length, length_count, dest);
+                 });
+
+  return ids;
+}
+
+auto SourceGen::GetSingleLengthIds(int length, int number)
+    -> llvm::ArrayRef<llvm::StringRef> {
+  llvm::SmallVector<llvm::StringRef>& ids =
+      ids_by_length.Insert(length, {}).value();
+
+  if (static_cast<int>(ids.size()) < number) {
+    ids.reserve(number);
+    for ([[maybe_unused]] int i : llvm::seq<int>(ids.size(), number)) {
```

Comment:
```suggestion
    for (int _ : llvm::seq<int>(ids.size(), number)) {
```
Try C++23 Today! :)

(Clang suppresses the `-Wunused` warning for variables named `_` in all language modes in preparation for C++23's special variables-named-`_` feature.)

There are a few more instances elsewhere in the PR if you're OK with pre-adopting this.